File: Medium/105/105.py
# @Date    : 11:25 03/25/2021
# @Author  : ClassicalPi
# @FileName: 105.py
# @Software: PyCharm
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def buildTree(self, preorder: list, inorder: list) -> TreeNode:
        dic={}
        for index,val in enumerate(inorder):
            dic.setdefault(val,index)
        def helper(start:int,end:int)->TreeNode or None:
            if start>=end:
                return None
            root_val=preorder.pop(0)
            root_index=dic[root_val]
            root=TreeNode(root_val,helper(start,root_index),helper(root_index+1,end))
            return root
        return helper(0,len(inorder))

File: GoVersion/Medium/105.go
package main

type TreeNode struct {
     Val int
     Left *TreeNode
     Right *TreeNode
}

func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder)==0 || len(inorder)==0{
		return nil
	}
	val:=preorder[0]
	index:=find(val,inorder)
	left:=buildTree(preorder[1:index+1],inorder[:index])
	right:=buildTree(preorder[index+1:],inorder[index+1:])
	return &TreeNode{
		Val:   val,
		Left:  left,
		Right: right,
	}
}

func find(target int, nums []int) int {
	for i:=0;i<len(nums);i+=1{
		if nums[i]==target {
			return i
		}
	}
	return -1
}